[feat] 공유 목표 조회 기능 추가(인기순 정렬 및 기간 필터 기능 추가) | Feature/byeongtaek

byeongtaek12 opened the pull request:
## 📌 관련 이슈
- Closes #89 

## ✨ 변경 사항
- 인기순 정렬 기능 추가
- 기간 필터 기능 추가

## 📷 Postman 
- 
![00000](https://github.com/user-attachments/assets/30ac3df1-6809-4801-9027-cc97950b155a)
![000001](https://github.com/user-attachments/assets/6814ebfe-eae1-4690-99cb-99535863611b)
![0701 1](https://github.com/user-attachments/assets/ede48475-b447-4a2e-871c-c452825450ef)
![010105050](https://github.com/user-attachments/assets/2b6b6939-24b6-403c-995e-12621f9ca50f)


## ✅ 체크리스트
- [ ] 관련 이슈에 연결됨
- [ ] 기능이 정상적으로 동작함
- [ ] 코드 리뷰를 받았음
- [ ] 불필요한 로그/코드 제거됨

## 💬 기타 참고 사항
- 테스트 방법, 추가 설명 등 필요한 경우 작성해 주세요.

byeongtaek12 committed:
[feat] : 에러 코드 추가 (#89)
1. 기간 설정을 연, 월, 일주일, 일 4가지 밖에 없기 때문에 나머지의 값이 들어오면 예외처리를 해줘야 하므로 추가

byeongtaek12 committed:
[refactor] : Goal Entity 리팩토링
1. commentCount와 likeCount 둘 다 더 큰 값이 들어올 수도 있기 때문에 int -> long 타입으로 수정
2. @CollectionTable 안에 post_tags와 post_id -> goal_tags와 goal_id로 수정

byeongtaek12 committed:
[feat] : 공유 목표 조회에 인기순 정렬 및 기간 필터 추가 (#89)
1. ShareGoalController 클래스 매개변수 추가
2. ShareGoalService 인터페이스 매개변수 추가
3. ShareGoalServiceImpl 클래스 매개변수 및 로직 추가
4. ShareGoalResponseDto 클래스 필드와 정적 팩토리 메서드에 필드 초기화 해주는 부분 추가
5. CustomShareGoalRepository 인터페이스 매개변수 추가
6. CustomShareGoalRepositoryImpl 클래스 매개변수 추가 및 로직 추가
7. ShareGoalServiceImplTest 단위테스트 추가
8. StringToPeriodConverter 클래스 구현
9. StringToSortConverter 클래스 구현
10. Sort Enum 클래스 구현
기능 - 인기순으로 정렬 및 기간 필터 추가

rlqor9688 commented on src/main/java/org/example/lifechart/domain/shareGoal/controller/ShareGoalController.java:
```diff
@@ -46,10 +48,13 @@ public ResponseEntity<ApiResponse<ShareGoalCursorResponseDto>> getShareGoals(
 		@RequestParam(required = false) Category category,
-		@RequestParam(required = false) Share share
+		@RequestParam(required = false) Share share,
+		@RequestParam(required = false)Sort sort,
```
Sort 앞에 띄어쓰기 하나 해주면 좋을 것 같네요!ㅋㅋ

byeongtaek12 replied:
바로 반영하겠습니다.

rlqor9688 commented on src/main/java/org/example/lifechart/domain/shareGoal/service/ShareGoalServiceImpl.java:
```diff
@@ -33,13 +38,27 @@ public class ShareGoalServiceImpl implements ShareGoalService {
 	@Transactional
```
@Transactional(readOnly=true)를 붙여주면 읽기 전용 메서드임을 명확히 보여줄 수 있고 성능 면에서도 유리할 것 같습니다!

rlqor9688 commented on src/main/java/org/example/lifechart/domain/shareGoal/service/ShareGoalServiceImpl.java:
공통적으로 조회 메서드에 @Transactional(readOnly = true)를 써주면 좋을 것 같습니다!

rlqor9688 reviewed (approved):
자잘한 코멘트 남겼습니다! 고생하셨습니다~

byeongtaek12 committed:
[refactor] : 공백 추가 및 Transactional(readOnly = true) 추가 (#89)
1. 피드백을 받아서 GET과 같은 조회메서드에는 flush나 변경감지 등이 필요없어서 readOnly를 붙였을 때 성능이 더 좋다라고 하여 추가

byeongtaek12 committed:
Merge branch 'dev' into feature/byeongtaek

byeongtaek12 commented:
> 자잘한 코멘트 남겼습니다! 고생하셨습니다~

모두 반영완료했습니다.✔

byeongtaek12 committed:
Merge branch 'dev' into feature/byeongtaek

ChangHoonS commented on src/main/java/org/example/lifechart/domain/shareGoal/controller/StringToSortConverter.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class StringToSortConverter implements Converter<String, Sort> {
```
컨버터로 변환하는 과정에서도 예외가 발생할 수 있다고 하는데 이 부분도 CustomException 사용하는 것도 좋아 보입니다.
참고 부탁드립니다.

byeongtaek12 replied:
예외 처리 추가 완료했습니다.

ChangHoonS commented on src/main/java/org/example/lifechart/domain/shareGoal/repository/CustomShareGoalRepositoryImpl.java:
```diff
@@ -22,27 +28,48 @@ public class CustomShareGoalRepositoryImpl implements CustomShareGoalRepository
+			LocalDateTime term = LocalDateTime.now().minus(period);
+			booleanBuilder.and(goal.createdAt.goe(term));
+		}
 		booleanBuilder.and(goal.status.eq(Status.ACTIVE));
```
booleanBuilder.and(goal.status.eq(Status.ACTIVE));
이렇게 자주 쓰이는 경우 메서드로 분리하면 재사용성이 증가하여 좋다고 합니다. 바꿀 필요는 없어 보이고 참고만 하시면 좋을 것 같습니다.

byeongtaek12 replied:
이 부분은 일단 참고하겠습니다.

ChangHoonS commented on src/test/java/org/example/lifechart/domain/shareGoal/service/ShareGoalServiceImplTest.java:
```diff
@@ -88,7 +92,12 @@ void setUp() {
+			.commentCount(5)
+			.likeCount(5)
 			.build();
+		Field createdAt = BaseEntity.class.getDeclaredField("createdAt");
```
이 부분에서 createdAt을 위에 private Field createdAt;으로 선언하고
메서드 상위에 createdAt = BaseEntity.class.getDeclaredField("createdAt"); 넣어주고
createdAt.setAccessible(true); 이 부분도 한번만 같이 선언해주면 중복을 피할 수 있다고 합니다.
참고로만 아시면 될 것 같습니다.

byeongtaek12 replied:
이 부분도 참고해놓겠습니다.

ChangHoonS reviewed (approved):
간단한 코멘트만 남겼고 승인해두겠습니다.

byeongtaek12 committed:
[refactor] : Converter 예외처리 추가
